chain/client/src/chunk_distribution_network.rs

```diff
@@ -0,0 +1,505 @@
+
+use borsh::BorshDeserialize;
+use near_async::messaging::Sender;
+use near_chain::{
+    blocks_delay_tracker::BlocksDelayTracker, chain::BlockMissingChunks,
+    orphan::OrphanMissingChunks,
+};
+use near_chain_configs::{ChunkDistributionNetworkConfig, ClientConfig};
+use near_chunks::adapter::ShardsManagerRequestFromClient;
+use near_primitives::{
+    hash::CryptoHash,
+    sharding::{PartialEncodedChunk, ShardChunkHeader},
+    static_clock::StaticClock,
+    types::{EpochId, ShardId},
+};
+use std::fmt;
+use tracing::error;
+
+/// Basic interface for the chunk distribution network.
+pub trait ChunkDistributionClient {
+    type Error;
+    type Response;
+
+    async fn lookup_chunk(
+        &self,
+        prev_hash: CryptoHash,
+        shard_id: ShardId,
+    ) -> Result<Option<PartialEncodedChunk>, Self::Error>;
+
+    async fn publish_chunk(
+        &mut self,
+        chunk: &PartialEncodedChunk,
+    ) -> Result<Self::Response, Self::Error>;
+}
+
+/// Helper struct for the Chunk Distribution Network Feature.
+#[derive(Debug, Clone)]
+pub struct ChunkDistributionNetwork {
+    pub client: reqwest::Client,
+    pub config: ChunkDistributionNetworkConfig,
+}
+
+impl ChunkDistributionNetwork {
+    pub fn new(config: &ClientConfig) -> Option<Self> {
+        config
+            .chunk_distribution_network
+            .as_ref()
+            .map(|c| Self { client: reqwest::Client::new(), config: c.clone() })
+    }
+}
+
+pub fn request_missing_chunks<C>(
+    blocks_missing_chunks: Vec<BlockMissingChunks>,
+    orphans_missing_chunks: Vec<OrphanMissingChunks>,
+    client: C,
+    blocks_delay_tracker: &mut BlocksDelayTracker,
+    shards_manager_adapter: &Sender<ShardsManagerRequestFromClient>,
+) where
+    C: ChunkDistributionClient + Clone + 'static,
+    C::Error: fmt::Debug,
+{
+    let now = StaticClock::utc();
+    for BlockMissingChunks { prev_hash, missing_chunks } in blocks_missing_chunks {
+        for chunk in missing_chunks {
+            blocks_delay_tracker.mark_chunk_requested(&chunk, now);
+            request_missing_chunk(client.clone(), chunk, shards_manager_adapter, prev_hash);
+        }
+    }
+
+    for OrphanMissingChunks { missing_chunks, epoch_id, ancestor_hash } in orphans_missing_chunks {
+        for chunk in missing_chunks {
+            blocks_delay_tracker.mark_chunk_requested(&chunk, now);
+            request_orphan_chunk(
+                client.clone(),
+                chunk,
+                shards_manager_adapter,
+                epoch_id.clone(),
+                ancestor_hash,
+            );
+        }
+    }
+}
+
+fn request_missing_chunk<C>(
+    client: C,
+    header: ShardChunkHeader,
+    adapter: &Sender<ShardsManagerRequestFromClient>,
+    prev_hash: CryptoHash,
+) where
+    C: ChunkDistributionClient + 'static,
+    C::Error: fmt::Debug,
+{
+    let shard_id = header.shard_id();
+    let thread_local_shards_manager_adapter = adapter.clone();
```

Comment:
nit: just take owned `adapter` and make it caller responsibility to call `.clone()`

Reply:
Addressed as part of https://github.com/near/nearcore/pull/10480/commits/2a4d18189d23cc4a4a8d72b248ac65b8d87006e0